### PR TITLE
C api for compact integer safe deserialization

### PR DESCRIPTION
**Pull request by mayeul-zama**

### CODE REVIEW

File: tfhe/src/shortint/server_key/mod.rs [hunk 1]
```diff
@@ -247,15 +247,23 @@ impl ShortintBootstrappingKey {
...
-    pub message_modulus: MessageModulus,
-    // Size of the carry buffer
-    pub carry_modulus: CarryModulus,
-    // Maximum number of operations that can be done before emptying the operation buffer
+    pub parameters: ShortintParameterSet,
     pub max_degree: MaxDegree,
-    // Modulus use for computations on the ciphertext
-    pub ciphertext_modulus: CiphertextModulus,
-    pub pbs_order: PBSOrder,
+}
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/mod.rs:252]: why this change ?

File: tfhe/src/shortint/server_key/compressed.rs [hunk 2]
```diff
@@ -48,14 +47,8 @@ impl ShortintCompressedBootstrappingKey {
...
-    // Size of the message buffer
-    pub message_modulus: MessageModulus,
-    // Size of the carry buffer
-    pub carry_modulus: CarryModulus,
-    // Maximum number of operations that can be done before emptying the operation buffer
+    pub parameters: ShortintParameterSet,
     pub max_degree: MaxDegree,
-    pub ciphertext_modulus: CiphertextModulus,
-    pub pbs_order: PBSOrder,
 }
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/compressed.rs:52]: same question

File: tfhe/c_api_tests/test_high_level_integers.c [hunk 3]
```diff
@@ -64,6 +64,66 @@ int uint8_public_key(const ClientKey *client_key, const PublicKey *public_key) {
...
+
+  ok = compact_fhe_uint8_list_try_encrypt_with_compact_public_key_u8(&lhs_clear, 1, public_key, &lhs);
+  assert(ok == 0);
+
+  ok = compact_fhe_uint8_list_safe_serialize(lhs, &value_buffer, 1 << 20);
+  assert(ok == 0);
+
+  deser_view.pointer = value_buffer.pointer;
+  deser_view.length = value_buffer.length;
+  ok = compact_fhe_uint8_list_safe_deserialize_conformant(deser_view, 1 << 20, server_key, &deserialized_lhs, 1);
```
Comment (tmontaigu) [inline tfhe/c_api_tests/test_high_level_integers.c:100]: For clarity, I think these `1 << 20` should be put in a `const` like `const uint64_t max_serialization_size = UINT64_C(1) << UINT64_C(20);`

File: tfhe/src/high_level_api/integers/mod.rs [hunk 4]
```diff
@@ -19,3 +19,71 @@ mod tests_signed;
...
+    ) -> Result<T, String>
+    where
+        T: Named + DeserializeOwned + ParameterSetConformant<ParameterSet = PBSParameters>,
+    {
+        let ps = sk
+            .integer_key
+            .key
+            .as_ref()
+            .unwrap()
+            .key
```
[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/src/high_level_api/integers/mod.rs]: Done

File: tfhe/src/shortint/server_key/mod.rs (hunk 1 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs]: Reverted

File: tfhe/src/shortint/server_key/compressed.rs (hunk 2 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/compressed.rs]: Reverted

**Comment by mayeul-zama** [comment]
@slab-ci cpu_fast_test

File: tfhe/c_api_tests/test_high_level_integers.c [hunk 5]
```diff
@@ -64,6 +64,68 @@ int uint8_public_key(const ClientKey *client_key, const PublicKey *public_key) {
   return ok;
 }
 
+int uint8_safe_serialization(const ClientKey *client_key, const ServerKey* server_key) {
```
Comment (tmontaigu) [inline tfhe/c_api_tests/test_high_level_integers.c:67]: I don't know if in the rust implementation there is a test, that checks that using 2 keys made with different (incompatible) params, if you mixmatch ciphertexts, the safe_deserialization properly returns that ciphertexts are not conformant ?

File: tfhe/c_api_tests/test_high_level_integers.c [hunk 6]
```diff
@@ -64,6 +64,68 @@ int uint8_public_key(const ClientKey *client_key, const PublicKey *public_key) {
...
+
+  struct CompactPublicKey *public_key;
+
+  ok = compact_public_key_new(deserialized_client_key, &public_key);
+  assert(ok == 0);
+
+  ok = compact_fhe_uint8_try_encrypt_with_compact_public_key_u8(lhs_clear, public_key, &lhs);
+  assert(ok == 0);
+
+  ok = compact_fhe_uint8_safe_serialize(lhs, &value_buffer,max_serialization_size);
```
Comment (IceTDrinker) [inline tfhe/c_api_tests/test_high_level_integers.c:97]: I know we don't have a C/C++ formatter for now, but missing space after comma

I think clang-format 14 is what is on my machine and it works reasonably well

File: tfhe/src/high_level_api/integers/mod.rs [hunk 7]
```diff
@@ -19,3 +19,62 @@ mod tests_signed;
...
+        reader: impl std::io::Read,
+        serialized_size_limit: u64,
+        sk: &ServerKey,
+    ) -> Result<T, String>
+    where
+        T: Named
+            + DeserializeOwned
+            + ParameterSetConformant<ParameterSet = CiphertextConformanceParams>,
+    {
+        let ps = sk.integer_key.pbs_key().key.conformance_params();
```
Comment (IceTDrinker) [inline tfhe/src/high_level_api/integers/mod.rs:55]: ps ? what does this mean ?

File: tfhe/src/shortint/parameters/mod.rs [hunk 8]
```diff
@@ -115,6 +115,45 @@ impl ClassicPBSParameters {
...
+    #[cfg(feature = "integer")]
+    pub fn to_integer_conformance_param(
+        &self,
+        num_blocks_per_integer: usize,
+    ) -> crate::integer::parameters::CiphertextConformanceParams {
+        crate::integer::parameters::CiphertextConformanceParams {
+            shortint_params: self.to_shortint_conformance_param(),
+            num_blocks_per_integer,
+        }
+    }
```
Comment (IceTDrinker) [inline tfhe/src/shortint/parameters/mod.rs:156]: can we implement a from for the crate::integer::parameters::CiphertextConformanceParams from the shortint parameter ? I don't like inverting the knowledge requirement for the layers, shortint does not have to know about integer

File: tfhe/src/shortint/parameters/mod.rs [hunk 9]
```diff
@@ -272,57 +312,21 @@ impl PBSParameters {
         matches!(self, Self::MultiBitPBS(_))
     }
 
-    pub fn to_shortint_conformance_param(&self) -> CiphertextConformanceParams {
-        let (pbs_order, expected_dim, message_modulus, ciphertext_modulus, carry_modulus);
+    #[cfg(feature = "integer")]
+    pub fn to_integer_conformance_param(
```
Comment (IceTDrinker) [inline tfhe/src/shortint/parameters/mod.rs:316]: same thing

File: tfhe/src/shortint/parameters/mod.rs [hunk 10]
```diff
@@ -501,11 +505,17 @@ impl ShortintParameterSet {
...
             ShortintParameterSetInner::PBSOnly(params) => params.encryption_key_choice(),
             ShortintParameterSetInner::WopbsOnly(params) => params.encryption_key_choice,
-            ShortintParameterSetInner::PBSAndWopbs(params, _) => params.encryption_key_choice(),
+            ShortintParameterSetInner::PBSAndWopbs(pbs_params, wop_pbs_params) => {
+                assert_eq!(
+                    pbs_params.encryption_key_choice(),
+                    wop_pbs_params.encryption_key_choice
+                );
+                pbs_params.encryption_key_choice()
+            }
```
Comment (IceTDrinker) [inline tfhe/src/shortint/parameters/mod.rs:518]: is this required ? If not I'd rather have it in a separate PR/issue as we keyswitch between those parameter sets and the parameter set we use for encryption is the one from the PBS, if you think this deserves a comment I'm fine with it but I don't think this change is correct (because we keyswitch from PBS to WOPBS and then back to PBS when required)

File: tfhe/src/shortint/server_key/mod.rs [hunk 11]
```diff
@@ -258,6 +258,28 @@ pub struct ServerKey {
...
+        let ct_params = LweCiphertextParameters {
+            lwe_dim,
+            ct_modulus: self.ciphertext_modulus,
+        };
+
+        CiphertextConformanceParams {
+            ct_params,
+            message_modulus: self.message_modulus,
+            carry_modulus: self.carry_modulus,
+            degree: Degree(self.message_modulus.0 - 1),
```
Comment (IceTDrinker) [inline tfhe/src/shortint/server_key/mod.rs:277]: do we have an issue to plan to support a range of valid degree ?

File: tfhe/src/integer/parameters/mod.rs [hunk 12]
```diff
@@ -117,3 +119,44 @@ pub const PARAM_MESSAGE_1_CARRY_1_KS_PBS_32_BITS: WopbsParameters = WopbsParamet
     ciphertext_modulus: CiphertextModulus::new_native(),
     encryption_key_choice: EncryptionKeyChoice::Big,
 };
+
+pub struct CiphertextConformanceParams {
+    pub shortint_params: shortint::parameters::CiphertextConformanceParams,
+    pub num_blocks_per_integer: usize,
+}
```
Comment (IceTDrinker) [inline tfhe/src/integer/parameters/mod.rs:126]: Down the line those are radix conformance params, but we will want to have the distinction vs. CRT, if you can have an issue about it and put those in the radix ciphertext module ?

File: tfhe/src/integer/parameters/mod.rs [hunk 13]
```diff
@@ -117,3 +119,44 @@ pub const PARAM_MESSAGE_1_CARRY_1_KS_PBS_32_BITS: WopbsParameters = WopbsParamet
...
+            num_blocks_per_integer: self.num_blocks_per_integer,
+            num_integers_constraint: list_constraint,
+        }
+    }
+}
+
+/// Structure to store the expected properties of a ciphertext list
+/// Can be used on a server to check if client inputs are well formed
+/// before running a computation on them
+pub struct CiphertextListConformanceParams {
```
Comment (IceTDrinker) [inline tfhe/src/integer/parameters/mod.rs:147]: If we can have the Compact naming if those are destined to be used with compact ciphertexts please (same for all structs related to compact ciphertexts)

File: tfhe/src/high_level_api/integers/mod.rs (hunk 7 above)
Comment (mayeul-zama) [reply tfhe/src/high_level_api/integers/mod.rs:55]: Renamed `parameter_set`

File: tfhe/src/shortint/parameters/mod.rs (hunk 8 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/parameters/mod.rs:156]: Sure, I did it this way because it's shorter to call but I agree it's not great
I'll change it

Comment (IceTDrinker) [reply tfhe/src/shortint/parameters/mod.rs:156]: I understand and it's good to make progress :) but yeah not mergeable I believe

File: tfhe/src/shortint/server_key/mod.rs (hunk 11 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/server_key/mod.rs:277]: Yes https://github.com/zama-ai/tfhe-rs-internal/issues/270

File: tfhe/src/integer/parameters/mod.rs (hunk 12 above)
Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs:126]: Sure, I'll move it
What would the issue contain?

File: tfhe/src/integer/parameters/mod.rs (hunk 13 above)
Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs:147]: It don´t think there's anything specific to compact (but to radix yes)

File: tfhe/src/shortint/server_key/mod.rs (hunk 11 above)
Comment (IceTDrinker) [reply tfhe/src/shortint/server_key/mod.rs:277]: perfect thanks

File: tfhe/src/integer/parameters/mod.rs (hunk 12 above)
Comment (IceTDrinker) [reply tfhe/src/integer/parameters/mod.rs:126]: Indicate that Crt Ciphertexts are different compared to Radix Ciphertexts and will required their own handling down the line

File: tfhe/src/integer/parameters/mod.rs (hunk 13 above)
Comment (IceTDrinker) [reply tfhe/src/integer/parameters/mod.rs:147]: aren't those used for the compact ciphertexts ? I may I have misread I just don't recall CiphertextLists other than compact ones at the integer level 🤔

File: tfhe/src/integer/parameters/mod.rs (hunk 12 above)
Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs:126]: There's no `ciphertext::radix` module
I'll rename them to `Radix___`

File: tfhe/src/integer/parameters/mod.rs (hunk 13 above)
Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs:147]: Indeed, we don't have any other lists (I assume we did)
But there is nothing specific to compact in it anyway

Comment (IceTDrinker) [reply tfhe/src/integer/parameters/mod.rs:147]: Ok, I still would like to have the compact naming as for now it corresponds to the compact part and it must be checking the compact LWE ciphertext list right ? This is not the expanded version ?

Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs:147]: It is currently only used for Compact
I'll rename it :)

Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs]: Done

File: tfhe/src/integer/parameters/mod.rs (hunk 12 above)
Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs]: Renamed

File: tfhe/src/shortint/parameters/mod.rs (hunk 10 above)
Comment (mayeul-zama) [reply tfhe/src/shortint/parameters/mod.rs]: Removed

File: tfhe/c_api_tests/test_high_level_integers.c (hunk 6 above)
Comment (mayeul-zama) [reply tfhe/c_api_tests/test_high_level_integers.c]: Done

**Comment by mayeul-zama** [comment]
@slab-ci cpu_fast_test

File: tfhe/src/integer/parameters/mod.rs [hunk 14]
```diff
@@ -117,3 +122,62 @@ pub const PARAM_MESSAGE_1_CARRY_1_KS_PBS_32_BITS: WopbsParameters = WopbsParamet
...
+    pub fn from_pbs_parameters(
+        pbs_parameters: &PBSParameters,
+        num_blocks_per_integer: usize,
+    ) -> RadixCiphertextConformanceParams {
+        RadixCiphertextConformanceParams {
+            shortint_params: pbs_parameters.to_shortint_conformance_param(),
+            num_blocks_per_integer,
+        }
+    }
+}
```
Comment (IceTDrinker) [inline tfhe/src/integer/parameters/mod.rs:165]: this could be only

`pub fn from_pbs_parameters<P: Into<PBSParameters>>(params: P, ...) { let params: PBSParameters = params.into();...}`

normally as PBSParameters is the "base enum" for ClassicPBSParameters and MultiBitPBS etc.

**Review by IceTDrinker — COMMENTED** [review]
Small comments but I think we should be close to merging

File: tfhe/src/high_level_api/integers/types/compact.rs [hunk 15]
```diff
@@ -95,13 +97,11 @@ where
...
 impl<P: IntegerParameter> ParameterSetConformant for GenericCompactInteger<P> {
-    type ParameterSet = PBSParameters;
-    fn is_conformant(&self, pbs_parameters: &PBSParameters) -> bool {
-        let params = CiphertextListConformanceParams {
-            shortint_params: pbs_parameters.to_shortint_conformance_param(),
-            num_blocks_per_integer: P::num_blocks(),
-            num_integers_constraint: ListSizeConstraint::exact_size(1),
-        };
+    type ParameterSet = RadixCiphertextConformanceParams;
+    fn is_conformant(&self, shortint_params: &RadixCiphertextConformanceParams) -> bool {
```
[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/src/high_level_api/integers/types/compact.rs:101]: No, I have an unpushed commit to rename :)

File: tfhe/src/high_level_api/integers/mod.rs [hunk 16]
```diff
@@ -19,3 +19,67 @@ mod tests_signed;
...
+    use crate::integer::parameters::RadixCiphertextConformanceParams;
+    use crate::named::Named;
+    use crate::shortint::parameters::CiphertextConformanceParams;
+    use crate::ServerKey;
+    use serde::de::DeserializeOwned;
+    use serde::Serialize;
+
+    pub fn safe_serialize<T>(
+        a: &T,
+        writer: impl std::io::Write,
```
[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/src/high_level_api/integers/mod.rs:37]: I don't think so
If a type `T` `impl Write`, the caller can pass a `&mut T` which also `impl Write` (https://doc.rust-lang.org/std/io/trait.Write.html#impl-Write-for-%26mut+W)

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/high_level_api/integers/mod.rs:37]: perfect

File: tfhe/src/high_level_api/integers/types/compact.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (mayeul-zama) [reply tfhe/src/high_level_api/integers/types/compact.rs]: Done

File: tfhe/src/integer/parameters/mod.rs (hunk 14 above)
Comment (mayeul-zama) [reply tfhe/src/integer/parameters/mod.rs]: Done

**Review by IceTDrinker — APPROVED** [review]
I think we are good to go for this first iteration that goes up to the HL API and C API, thanks a lot for you hard work ! 🙂

**Review by IceTDrinker — APPROVED** [review]
Thanks !

File: tfhe/c_api_tests/test_high_level_integers.c (hunk 5 above)
Comment (tmontaigu) [reply tfhe/c_api_tests/test_high_level_integers.c]: Ok 👍

**Comment by IceTDrinker** [comment]
@slab-ci csprng_randomness_testing

**Comment by tmontaigu** [comment]
@slab-ci csprng_randomness_testing
